Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -289,16 +289,20 @@ private class BetterHTTPParser {
             // does not meet the requirement of RFC 7230. This is an outstanding http_parser issue:
             // https://github.com/nodejs/http-parser/issues/251. As a result, we check for these status
             // codes and override http_parser's handling as well.
-            guard let method = self.requestHeads.popFirst()?.method else {
-                self.richerError = NIOHTTPDecoderError.unsolicitedResponse
-                return .error(HPE_UNKNOWN)
-            }
-
-            if method == .HEAD || method == .CONNECT {
-                skipBody = true
-            } else if statusCode / 100 == 1 ||  // 1XX codes
-                statusCode == 204 || statusCode == 304 {
+            
+            if statusCode / 100 == 1 && statusCode != 101 {
                 skipBody = true
+            } else {
+                guard let method = self.requestHeads.popFirst()?.method else {
+                    self.richerError = NIOHTTPDecoderError.unsolicitedResponse
```

Comment:
we should also add some actual unit tests (using `EmbeddedChannel` and not an actual TCP connection) testing all of the new functionality for both client and server. These tests are usually quite straightforward to write, for example: https://github.com/apple/swift-nio/blob/master/Tests/NIOHTTP1Tests/HTTPDecoderTest.swift#L549

For example for the request decoder (client) part:

```swift
        [...]
        var x = channel.allocator.buffer(capacity: 128)
        x.writeStaticString("x")
        var buffer = channel.allocator.buffer(capacity: 128)
        buffer.writeStaticString("HTTP/1.1 100 continue\r\n\r\n HTTP/1.1 100 continue\r\n\r\n HTTP/1.1 200 ok\r\ncontent-length: 1\r\n\r\nx")

        XCTAssertNoThrow(try channel.pipeline.addHandler(ByteToMessageHandler(HTTPResponseDecoder(leftOverBytesStrategy: .fireError))).wait())
        XCTAssertNoThrow(try channel.writeOutbound(HTTPClientRequestPart.head(.init(version: .init(major: 1, minor: 1),
                                                                                    method: .GET, uri: "/"))))
        XCTAssertNoThrow(try channel.writeInbound(buffer))
        XCTAssertNoThrow(XCTAssertEqual(HTTPClientResponsePart.head(.init(version: .init(major: 1, minor: 1),
                                                                          status: .continue)), try channel.readInbound()))
        XCTAssertNoThrow(XCTAssertEqual(HTTPClientResponsePart.end(nil), try channel.readInbound()))
        XCTAssertNoThrow(XCTAssertEqual(HTTPClientResponsePart.head(.init(version: .init(major: 1, minor: 1),
                                                                          status: .continue)), try channel.readInbound()))
        XCTAssertNoThrow(XCTAssertEqual(HTTPClientResponsePart.end(nil), try channel.readInbound()))
        XCTAssertNoThrow(XCTAssertEqual(HTTPClientResponsePart.head(.init(version: .init(major: 1, minor: 1),
                                                                          status: .ok)), try channel.readInbound()))
        XCTAssertNoThrow(XCTAssertEqual(HTTPClientResponsePart.body(x), try channel.readInbound()))
        XCTAssertNoThrow(XCTAssertEqual(HTTPClientResponsePart.end(nil), try channel.readInbound()))
        [...]
```